File: NEsper.Tests/pattern/observer/TestTimerCronObserver.cs
///////////////////////////////////////////////////////////////////////////////////////
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
///////////////////////////////////////////////////////////////////////////////////////

using System;

using NUnit.Framework;

using net.esper.core;
using net.esper.compat;
using net.esper.pattern;
using net.esper.schedule;
using net.esper.support.events;
using net.esper.support.guard;
using net.esper.support.schedule;
using net.esper.support.view;

namespace net.esper.pattern.observer
{
	[TestFixture]
	public class TestTimerCronObserver
	{
	    private TimerAtObserver observer;
	    private SchedulingServiceImpl scheduleService;
	    private SupportObserverEvaluator evaluator;
	    private MatchedEventMap beginState;

	    [SetUp]
	    public void SetUp()
	    {
	        beginState = new MatchedEventMapImpl();

	        scheduleService = new SchedulingServiceImpl();
	        StatementContext stmtContext = SupportStatementContextFactory.MakeContext(scheduleService);
	        PatternContext context = new PatternContext(stmtContext, 1, null);

	        ScheduleSpec scheduleSpec = new ScheduleSpec();
	        scheduleSpec.AddValue(ScheduleUnit.SECONDS, 1);

	        evaluator = new SupportObserverEvaluator();

	        observer =  new TimerAtObserver(scheduleSpec, context, beginState, evaluator);
	    }

	    [Test]
	    public void testStartAndObserve()
	    {
	        scheduleService.Time = (0);
	        observer.StartObserve();
	        scheduleService.Time = (1000);
	        SupportSchedulingServiceImpl.EvaluateSchedule(scheduleService);
	        Assert.AreEqual(beginState, evaluator.GetAndClearMatchEvents()[0]);

	        // Test start again
	        observer.StartObserve();
	        scheduleService.Time = (60999);
	        SupportSchedulingServiceImpl.EvaluateSchedule(scheduleService);
	        Assert.AreEqual(0, evaluator.GetMatchEvents().Count);

	        scheduleService.Time = (61000); // 1 minute plus 1 second
	        SupportSchedulingServiceImpl.EvaluateSchedule(scheduleService);
	        Assert.AreEqual(beginState, evaluator.GetAndClearMatchEvents()[0]);
	    }

	    [Test]
	    public void testStartAndStop()
	    {
	        // Start then stop
	        scheduleService.Time = (0);
	        observer.StartObserve();
	        observer.StopObserve();
	        scheduleService.Time = (1000);
	        SupportSchedulingServiceImpl.EvaluateSchedule(scheduleService);
	        Assert.AreEqual(0, evaluator.GetAndClearMatchEvents().Count);

	        // Test start again
	        observer.StartObserve();
	        scheduleService.Time = (61000);
	        SupportSchedulingServiceImpl.EvaluateSchedule(scheduleService);
	        Assert.AreEqual(beginState, evaluator.GetAndClearMatchEvents()[0]);

	        observer.StopObserve();
	        observer.StartObserve();

	        scheduleService.Time = (150000);
	        SupportSchedulingServiceImpl.EvaluateSchedule(scheduleService);
	        Assert.AreEqual(beginState, evaluator.GetAndClearMatchEvents()[0]);
	    }

	    [Test]
	    public void testInvalid()
	    {
	        try
	        {
	            observer.StartObserve();
	            observer.StartObserve();
	            Assert.Fail();
	        }
	        catch (IllegalStateException ex)
	        {
	            // Expected exception
	        }
	    }

	}
} // End of namespace

File: trunk/esper/src/test/java/net/esper/pattern/observer/TestTimerCronObserver.java
package net.esper.pattern.observer;

import junit.framework.TestCase;
import net.esper.pattern.MatchedEventMap;
import net.esper.pattern.PatternContext;
import net.esper.pattern.MatchedEventMapImpl;
import net.esper.schedule.ScheduleSpec;
import net.esper.type.ScheduleUnit;
import net.esper.schedule.SchedulingServiceImpl;
import net.esper.support.guard.SupportObserverEvaluator;
import net.esper.support.schedule.SupportSchedulingServiceImpl;
import net.esper.support.view.SupportStatementContextFactory;
import net.esper.core.StatementContext;

public class TestTimerCronObserver extends TestCase
{
    private TimerAtObserver observer;
    private SchedulingServiceImpl scheduleService;
    private SupportObserverEvaluator evaluator;
    private MatchedEventMap beginState;

    public void setUp()
    {
        beginState = new MatchedEventMapImpl();

        scheduleService = new SchedulingServiceImpl();
        StatementContext stmtContext = SupportStatementContextFactory.makeContext(scheduleService);
        PatternContext context = new PatternContext(stmtContext, 1, null);

        ScheduleSpec scheduleSpec = new ScheduleSpec();
        scheduleSpec.addValue(ScheduleUnit.SECONDS, 1);

        evaluator = new SupportObserverEvaluator();

        observer =  new TimerAtObserver(scheduleSpec, context, beginState, evaluator);
    }

    public void testStartAndObserve()
    {
        scheduleService.setTime(0);
        observer.startObserve();
        scheduleService.setTime(1000);
        SupportSchedulingServiceImpl.evaluateSchedule(scheduleService);
        assertEquals(beginState, evaluator.getAndClearMatchEvents().get(0));

        // Test start again
        observer.startObserve();
        scheduleService.setTime(60999);
        SupportSchedulingServiceImpl.evaluateSchedule(scheduleService);
        assertEquals(0, evaluator.getMatchEvents().size());

        scheduleService.setTime(61000); // 1 minute plus 1 second
        SupportSchedulingServiceImpl.evaluateSchedule(scheduleService);
        assertEquals(beginState, evaluator.getAndClearMatchEvents().get(0));
    }

    public void testStartAndStop()
    {
        // Start then stop
        scheduleService.setTime(0);
        observer.startObserve();
        observer.stopObserve();
        scheduleService.setTime(1000);
        SupportSchedulingServiceImpl.evaluateSchedule(scheduleService);
        assertEquals(0, evaluator.getAndClearMatchEvents().size());

        // Test start again
        observer.startObserve();
        scheduleService.setTime(61000);
        SupportSchedulingServiceImpl.evaluateSchedule(scheduleService);
        assertEquals(beginState, evaluator.getAndClearMatchEvents().get(0));

        observer.stopObserve();
        observer.startObserve();

        scheduleService.setTime(150000);
        SupportSchedulingServiceImpl.evaluateSchedule(scheduleService);
        assertEquals(beginState, evaluator.getAndClearMatchEvents().get(0));
    }

    public void testInvalid()
    {
        try
        {
            observer.startObserve();
            observer.startObserve();
            fail();
        }
        catch (IllegalStateException ex)
        {
            // Expected exception
        }
    }

}
